### PR TITLE
Modularization: clientv3/snapshot -> etcdctl/snapshot

### PR DESCRIPTION
**Pull request by ptabor**
Preserved snapshot.Save functionality in clientv3 and migrated rest (SnapshotManager.Status, SnapshotManager.Restore) using:

-  `git mv clientv3/snapshit  etcdctl/`
-  fix imports 
- `go fmt ./...`

"snapshot" code was the only remaining dependency of client on 'server'
code. The code is solelly used by etcdctl. Long-term the snapshot code
should be migrated to 'etcdadm' style of tool such that we can
distinguish tool solelly depending on networking API vs. tools that
operation on etcd files directly.

### CODE REVIEW

**Comment by gyuho** [comment]
`clientv3/snapshot` can be used for external projects (e.g. download snapshot in Go code)?

**Comment by ptabor** [comment]
> `clientv3/snapshot` can be used for external projects (e.g. download snapshot in Go code)?

@gyuho Thank you for pointing that out. Indeed that SnapshotManager implemented 3 functionalities: 
    1. Save:  download the snapshot over the public API and dump it to file
    2. Status: that opened the file using bolt and mvcc to verify the content.
    3. Restore: That work directly on etcd files and depend on etcd-server implementation details.  

I refactored the code such that snapshot.Save is preserved withing clientv3 code -> as its clean from dependencies perspectivces

I think that is OK that users that need a deep insight into etcd internal files, will need to import a separate package: 
 (etcdctl for now, and I hope that sth like etcdadm in the future). 

WDYT ?

**Comment by ptabor** [comment]
@gyuho Could you, please, check whether the current code and explanation in https://github.com/etcd-io/etcd/pull/12369#issuecomment-705420655 addresses the concern.

**Comment by gyuho** [comment]
> "snapshot" code was the only remaining dependency of client on 'server' code.

Oh, now I see your point.

Thanks for driving this.

**Review by gyuho — APPROVED** [review]

**Comment by gyuho** [comment]
Can we highlight this as a breaking change in 3.5?

(can be done in a separate PR)

**Comment by ptabor** [comment]
> Can we highlight this as a breaking change in 3.5?
> (can be done in a separate PR)

Done in the same PR. Thank you for review !

File: etcdctl/snapshot/v3_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+	dbClosed = true
+
+	commit := len(s.cl.Members())
+
+	// update consistentIndex so applies go through on etcdserver despite
+	// having a new raft instance
+	be := backend.NewDefaultBackend(dbpath)
+
+	ci := cindex.NewConsistentIndex(be.BatchTx())
+	ci.SetConsistentIndex(uint64(commit))
```
Comment (ahrtr) [inline etcdctl/snapshot/v3_snapshot.go:359]: This reset the consistent_index to `len(s.cl.Members())`. It's a little weird,  but the functionality isn't broken.

I think the consistent_index should keep it as it's, and update the commit index to `existing_consistent_index + len(s.cl.Members())`. see also https://github.com/etcd-io/etcd/pull/12369#discussion_r3873399978

File: etcdctl/snapshot/v3_snapshot.go [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+		}
+		ents[i] = raftpb.Entry{
+			Type:  raftpb.EntryConfChange,
+			Term:  1,
+			Index: uint64(i + 1),
+			Data:  d,
+		}
+	}
+
+	commit, term := uint64(len(ents)), uint64(1)
```
Comment (ahrtr) [inline etcdctl/snapshot/v3_snapshot.go:443]: The commit index should be `existing_consistent_index + len(s.cl.Members())`.

**Pull request merged**